Slice through ODF fields

Garyfallidis opened the pull request:
This PR introduces a new viz.actor (odf_slicer) that allows to slice ODF fields in native or world coordinates  matching the underlying anatomy. So, it works nicely with the viz.actor.slicer.

Some of the new features include applying a colormap using all the voxels in the slice or volume. 
![jet_odfs](https://cloud.githubusercontent.com/assets/134276/19972774/7e15bc08-a1b9-11e6-9e04-034766f3944f.png)

Also all the previous capabilities found in sphere_func are available too. Finally, the new function works with memory mapped ODFs and masking which helps reduce memory and video memory usage. I prefer to wait for the UI elements PR (https://github.com/nipy/dipy/pull/1140)  for creating a tutorial for this function.  With the new UI elements we can nicely upgrade some of our reconstruction and viz tutorials. But in the meantime you can review this PR with the tests.

MarcCote commented on dipy/viz/actor.py:
```diff
@@ -539,3 +540,186 @@ def axes(scale=(1, 1, 1), colorx=(1, 0, 0), colory=(0, 1, 0), colorz=(0, 0, 1),
+def odf_slicer(odfs, affine=None, mask=None, sphere=None, scale=2.2,
+               norm=True, radial_scale=True, opacity=1.,
+               colormap=None, global_cm=False):
+    """ Slice spherical fields in native or wold coordinates
```
wold => world

MarcCote commented on dipy/viz/actor.py:
```diff
@@ -539,3 +540,186 @@ def axes(scale=(1, 1, 1), colorx=(1, 0, 0), colory=(0, 1, 0), colorz=(0, 0, 1),
+            cols = np.ascontiguousarray(
+                np.reshape(cols, (cols.shape[0] * cols.shape[1],
+                           cols.shape[2])), dtype='f4')
+        # cols = np.interp(cols, [0, 1], [0, 255]).astype('ubyte')
```
unused

MarcCote commented on dipy/viz/actor.py:
```diff
@@ -539,3 +540,186 @@ def axes(scale=(1, 1, 1), colorx=(1, 0, 0), colory=(0, 1, 0), colorz=(0, 0, 1),
+                np.reshape(cols, (cols.shape[0] * cols.shape[1],
+                           cols.shape[2])), dtype='f4')
+        # cols = np.interp(cols, [0, 1], [0, 255]).astype('ubyte')
+        # vtk_colors = numpy_to_vtk_colors(255 * cols)
```
unused

Garyfallidis commented:
I believe we should have a PR soon to rename all test_fvtk** and a PR to update all tutorials to not be using fvtk. After we get in the UI elements we should do that. Maybe Ranveer can take the lead on that RF. But there is no need to do this RF in this PR. If you agree.

MarcCote commented:
Should also have a look at https://github.com/nipy/dipy/pull/944

arokem commented on dipy/viz/colormap.py:
```diff
@@ -255,3 +261,52 @@ def line_colors(streamlines, cmap='rgb_standard'):
+lowercase_cm_name = {'blues': 'Blues', 'accent': 'Accent'}
+
+
+def create_colormap(v, name='jet', auto=True):
```
"jet" is a really bad colormap (see e.g., ). Can we use switch to using a different default (viridis??)? If we want to keep this as a default to support back-compatibility, can we at least add a warning that we will change this in the future?

Garyfallidis replied:
Nope there is no reason to change the default behavior of this function neither to add another warning. What we need to do is use different colormaps in our tutorials. I will add something in the documentation. To suggest using viridis or inferno. Sounds good? By the way I don't think there is one colormap that is good for all the different jobs.

arokem replied:
I didn't say one colormap is good for all the different jobs you might want to do with it, just that this one is particularly bad for any task, and we shouldn't use it as the default (see: https://jakevdp.github.io/blog/2014/10/16/how-bad-is-your-colormap/). Some visualization software goes so far as to prevent users from using "jet": https://github.com/mwaskom/seaborn/commit/9f5a177639a7b5d5d8c7cce98def0e2e10c947b0

I don't think we need to go that far, but users will tend to stick to the default, so I think we should change it.

arokem replied:
I respectfully disagree. The colormap is supposed to help derive the information from the data, but this one hides and distorts information.

Garyfallidis replied:
Okay, so what should be the default? Ideas?

Garyfallidis replied:
From top to bottom. Inferno, jet and viridis on a response function. Clearly jet shows the minima better. But the other have a smoother transition.

![inferno](https://cloud.githubusercontent.com/assets/134276/19987833/75213fd4-a1f3-11e6-8455-487ad0488ca1.png)
![jet](https://cloud.githubusercontent.com/assets/134276/19987834/781f18dc-a1f3-11e6-8291-f28483852991.png)
![viridis](https://cloud.githubusercontent.com/assets/134276/19987836/7b36c826-a1f3-11e6-9045-7f9352b30b2b.png)

arokem commented on dipy/viz/actor.py:
```diff
@@ -539,3 +540,186 @@ def axes(scale=(1, 1, 1), colorx=(1, 0, 0), colory=(0, 1, 0), colorz=(0, 0, 1),
+    opacity : float
+        Takes values from 0 (fully transparent) to 1 (non-transparent)
+    colormap : None or str
+        If None then no color is used.
```
No color, or the default color ("jet"? See comment below)

arokem commented on dipy/viz/actor.py:
```diff
@@ -539,3 +540,186 @@ def axes(scale=(1, 1, 1), colorx=(1, 0, 0), colory=(0, 1, 0), colorz=(0, 0, 1),
+    opacity : float
+        Takes values from 0 (fully transparent) to 1 (non-transparent)
+    colormap : None or str
+        If None then no color is used.
```
No color, or the default colormap?

Garyfallidis replied:
No color for GL means all is white. Will add something.

arokem commented on dipy/viz/actor.py:
```diff
@@ -539,3 +540,186 @@ def axes(scale=(1, 1, 1), colorx=(1, 0, 0), colory=(0, 1, 0), colorz=(0, 0, 1),
+    cells = vtk.vtkCellArray()
+    cells.SetCells(ncells, all_faces_vtk)
+
+    if colormap is not None:
```
I see. It really skips this block if there's no colormap argument. So what does it look like then? 

BTW - could you share the code that created the example that you pasted into the PR text? I'd like to try running this on my machine to see that it works well (on Python 3!!)

Garyfallidis replied:
Just take the reconst_csd.py example or even better your SFM example and replace sphere_funcs with odf_slicer. Then to change the view just select the ROI you want with the display_extent method.

arokem commented on dipy/viz/colormap.py:
```diff
@@ -255,3 +261,52 @@ def line_colors(streamlines, cmap='rgb_standard'):
+    Notes
+    -----
+    Dipy supports a few colormaps for those who do not use Matplotlib, for
+    more colormaps consider downloading Matplotlib.
```
Maybe a link to the MPL website?

Garyfallidis replied:
ok

Garyfallidis commented on dipy/viz/actor.py:
```diff
@@ -539,3 +540,186 @@ def axes(scale=(1, 1, 1), colorx=(1, 0, 0), colory=(0, 1, 0), colorz=(0, 0, 1),
+    radial_scale : bool
+        Scale sphere points according to odf values.
+    opacity : float
+        Takes values from 0 (fully transparent) to 1 (non-transparent)
```
ok

arokem commented:
'viridis' or 'hot'

On Thu, Nov 3, 2016 at 2:44 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> *@Garyfallidis* commented on this pull request.
> ------------------------------
>
> In dipy/viz/colormap.py <https://github.com/nipy/dipy/pull/1143>:
>
> > @@ -255,3 +261,52 @@ def line_colors(streamlines, cmap='rgb_standard'):
>                      for streamline in streamlines]
>
>      return np.vstack(col_list)
> +
> +
> +lowercase_cm_name = {'blues': 'Blues', 'accent': 'Accent'}
> +
> +
> +def create_colormap(v, name='jet', auto=True):
>
> Okay, so what should be the default? Ideas?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1143>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNpZ5_mBcW49LaS2bwl-vXVVcq7dnks5q6lWogaJpZM4KoiF->
> .
>

arokem commented:
I'll amend that: I would strongly favor 'viridis' as a default.

It was designed through a careful process that took into account the best
knowledge we have of HCI perceptual considerations.


On Thu, Nov 3, 2016 at 3:32 PM, Ariel Rokem <arokem@gmail.com> wrote:

> 'viridis' or 'hot'
>
> On Thu, Nov 3, 2016 at 2:44 PM, Eleftherios Garyfallidis <
> notifications@github.com> wrote:
>
>> *@Garyfallidis* commented on this pull request.
>> ------------------------------
>>
>> In dipy/viz/colormap.py <https://github.com/nipy/dipy/pull/1143>:
>>
>> > @@ -255,3 +261,52 @@ def line_colors(streamlines, cmap='rgb_standard'):
>>                      for streamline in streamlines]
>>
>>      return np.vstack(col_list)
>> +
>> +
>> +lowercase_cm_name = {'blues': 'Blues', 'accent': 'Accent'}
>> +
>> +
>> +def create_colormap(v, name='jet', auto=True):
>>
>> Okay, so what should be the default? Ideas?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/nipy/dipy/pull/1143>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAHPNpZ5_mBcW49LaS2bwl-vXVVcq7dnks5q6lWogaJpZM4KoiF->
>> .
>>
>
>

Garyfallidis commented:
Did they consider spherical 3D functions? Or only images in their evaluation?

Garyfallidis commented:
Because they chose one that does not mean we need to chose the same. Here veridis is option 4 for example
https://bids.github.io/colormap/

So, here is magma.
![magma](https://cloud.githubusercontent.com/assets/134276/19988768/825cbe4e-a1f8-11e6-8c0e-912240b02cee.png)

To me veridis looks really depressing. Too green.

Why don't we all investigate a bit which colormap serves dMRI data better? It's good we started this discussion. But we do not need to make a decision immediately. Let's play with different colormaps and she which one prevails.

Garyfallidis commented:
Plasma looks good too.

![image](https://cloud.githubusercontent.com/assets/134276/19989461/0ac7f9fc-a1fd-11e6-8f28-f40b94bb7280.png)

arokem commented:
Yes: viridis, inferno, plasma, and magma are all equally good from a
perceptual/HCI point of view.

I agree that plasma looks good, and has nice cheerful colors. I'd be happy
to replace jet with plasma as a default.

On Thu, Nov 3, 2016 at 4:38 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Plasma looks good too.
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/134276/19989461/0ac7f9fc-a1fd-11e6-8f28-f40b94bb7280.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1143#issuecomment-258306234>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNuAwXbUZQ3CA8WjeYpN3ajCenM_vks5q6nB3gaJpZM4KoiF->
> .
>

MarcCote commented:
Yes, plasma looks great, we can go with that. For fun, can you post one with viridis, I'm curious!

arokem commented:
@MarcCote : there's one with viridis here: https://github.com/nipy/dipy/pull/1143#discussion_r86457402

Not too bad, in my opinion, but definitely tending towards the green shades

MarcCote commented:
On further consideration, the default for the colormap should be set in the `create_colormap` function and avoid setting it in every function that accepts a colormap in input. Whether or not we should do it in this PR is up to you.

Garyfallidis commented:
Hello @arokem and @MarcCote I added plasma as the default for the odf_slicer and added a message in create_colormap suggesting to users to prefer other colormaps. I think that is better than raising a warning at this stage. This direction gives us time to look into different colormaps but also informs people to start looking and trying alternatives.

arokem commented on dipy/viz/colormap.py:
```diff
@@ -255,3 +261,59 @@ def line_colors(streamlines, cmap='rgb_standard'):
+        msg = 'Jet is a popular colormap but can often be misleading and'
+        msg += 'we will remove it from being the default in the near future.'
+        msg += 'Try for example plasma, viridis, hot or inferno.'
+        print(msg)
```
I think the direction is right. We want to warn people of this now, and deprecate later. You probably want to raise a warning here, rather than just printing a message, though.

arokem replied:
It's a bit strange that the default behavior raises a warning. Why not set the default to be plasma here as well? I guess we do want the default to be consistent across different functions, and you changed it to plasma in several other places.

MarcCote commented:
@Garyfallidis having a closer look it seems that the other test with slicers `test_slicer` is being skip (because of `@npt.dec.skipif(skip_slicer)`) on Travis since it is still using VTK5.8.
See https://github.com/nipy/dipy/blob/master/dipy/viz/tests/test_fvtk_actors.py#L25

Garyfallidis commented:
Closing on behalf of https://github.com/nipy/dipy/pull/1269
